In [7]:
import scipy.io as sio
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import savgol_filter

Import time data

In [8]:
participant='01'
mat_data = sio.loadmat('D:\FYP\Datasets\dataset2\VP0'+participant+'-NIRS\mrk_nback')
my_data = mat_data['mrk_nback']['time'] # For a structure
print(my_data[0][0][0])
time=my_data[0][0][0]

[  46560  116160  185568  255168  324576  394176  463776  533280  602496
  750820  820324  890020  959524 1028836 1098244 1167748 1237348 1306948
 1456336 1525936 1595344 1664848 1734352 1803952 1873360 1942960 2012464]


Import fNIRS data

In [9]:

df=pd.read_csv('D:\FYP\Datasets\dataset2_csv\deoxy\deoxy_'+participant+'.csv')
#print(df)

In [10]:
new=np.array(time)
time_new=np.insert(new,0,0)
print(time_new)

[      0   46560  116160  185568  255168  324576  394176  463776  533280
  602496  750820  820324  890020  959524 1028836 1098244 1167748 1237348
 1306948 1456336 1525936 1595344 1664848 1734352 1803952 1873360 1942960
 2012464]


In [11]:

# Calculate the number of data points in each session
numDataPoints = np.diff(time_new) / 100  # Sampling rate is 10 Hz (adjust if needed)
numDataPoints = numDataPoints.astype(int) 
print(numDataPoints)

[ 465  696  694  696  694  696  696  695  692 1483  695  696  695  693
  694  695  696  696 1493  696  694  695  695  696  694  696  695]


First session of participant1 of all 36 channels

print(df.iloc[0:465,:])


In [12]:
val=[]
start_index=0
for t in numDataPoints:
    #print('session')
    end_index=start_index+t
    #print('start and end',start_index,end_index)
    newdf=df.iloc[start_index:end_index,:]
    start_index=end_index
    mean_values=[]
    for channel in range(36):
        x=newdf.iloc[:,channel]
        #print(len(x))
        #print(channel)
        y=savgol_filter(x, 121, 3)
        baseline_mean=np.mean(y)
        y=y-baseline_mean
        mean_values.append(np.mean(y))
        #print(np.mean(y))
    #print(mean_values)
    val.append(mean_values)
meandf=pd.DataFrame(val)
print(meandf)
filename='D:\FYP\Datasets\dataset2_features\mean_'+participant
meandf.to_csv(filename)

              0             1             2             3             4   \
0   1.492235e-18  1.193788e-19  1.298245e-18 -1.641459e-19 -1.585500e-20   
1  -1.276118e-18  0.000000e+00  7.178166e-19  0.000000e+00 -1.595148e-19   
2   1.279796e-18 -1.999681e-20 -1.879700e-18 -3.399458e-19  7.998725e-20   
3   6.380592e-19 -5.981805e-20 -6.380592e-19  0.000000e+00 -7.975740e-20   
4  -9.598470e-19  0.000000e+00  5.599107e-19 -3.999362e-20  3.999362e-20   
5   1.993935e-20  4.984838e-20  2.392722e-19 -5.981805e-20 -1.794542e-19   
6  -2.791509e-19  5.981805e-20  4.386657e-19  7.975740e-20  0.000000e+00   
7  -1.397763e-19 -7.488015e-20 -1.442691e-18  2.396165e-19  1.996804e-20   
8   1.082949e-18  1.002730e-20 -6.417474e-19 -1.804915e-19 -1.002730e-19   
9  -2.994533e-19  0.000000e+00  5.989066e-19  0.000000e+00  0.000000e+00   
10  1.597443e-19  3.194886e-19  7.987216e-19  7.987216e-20  1.597443e-19   
11  0.000000e+00 -6.480289e-20  0.000000e+00  0.000000e+00  3.987870e-20   
12  3.194886